packages/runtime/src/state/builder-api/proxy.ts

```diff
@@ -0,0 +1,45 @@
+import { type HostAction, isHostAction } from '../host-api'
+
+import { type BuilderAction, hasTransferables } from './actions'
+import { MessageChannel } from './message-channel'
+
+export class BuilderAPIProxy {
+  private messageChannel: MessageChannel
+
+  constructor({ appOrigin }: { appOrigin: string }) {
+    this.messageChannel = new MessageChannel({ appOrigin })
+  }
+
+  // low-level, action-based API
+
+  setup({ onHostAction }: { onHostAction: (action: HostAction) => void }): void {
+    this.messageChannel.setup((event: MessageEvent) => {
+      const action = event.data
+
+      if (!isHostAction(action)) {
+        console.warn('Unexpected host action', action)
+        return
+      }
+
+      onHostAction(action)
+    })
+  }
+
+  execute(action: BuilderAction): void {
+    if (hasTransferables(action)) {
+      const { transferables, ...forwardedAction } = action
```

Comment:
In particular, we don't have to worry about transferables, they are handled automatically behind the scenes.